### PR TITLE
MNMG Neighborhood Sampling

### PR DESCRIPTION
**Pull request by aschaffer**
This PR tracks work on MNMG Neighborhood Sampling, for G(C)NN needs.

Dependencies:
1. https://github.com/rapidsai/cugraph/pull/1982
2. https://github.com/rapidsai/cugraph/pull/2064
3. Integration of https://github.com/rapidsai/cugraph-ops/pull/24 into `cugraph`

### CODE REVIEW

**Comment by aschaffer** [comment]
rerun tests

File: cpp/src/sampling/nbr_sampling.cu
Comment (seunghwak) [inline cpp/src/sampling/nbr_sampling.cu:1]: Better rename this to `nbr_sampling_mg.cu` to follow our convention.

File: cpp/src/sampling/nbr_sampling.cuh
Comment (seunghwak) [inline cpp/src/sampling/nbr_sampling.cuh:475]: Should we really duplicate documentations (in the public header and here as well.... we should put extra-effort to keep this up-to-date and in-sync with the documentation in the public header).

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: The change is motivated by the requirement that the indices be returned too. Because there is a filtering happening in the sources, destinations, the same filtering must apply to the indices. This is a prim not exposable to API. The modification in place avoids an additional copy. This was authored by @kaatish , and we agreed to this format.

File: cpp/include/cugraph/algorithms.hpp
Comment (aschaffer) [reply cpp/include/cugraph/algorithms.hpp:1548]: Fixed that.

File: cpp/include/cugraph/algorithms.hpp
Comment (aschaffer) [reply cpp/include/cugraph/algorithms.hpp:1552]: Fixed that (default-true is `with` replacement, not `without`, that was a mistake in docs; fixed).

File: cpp/include/cugraph/algorithms.hpp
Comment (aschaffer) [reply cpp/include/cugraph/algorithms.hpp:1568]: This is targeting a very _specific_ implementation _with_ MPI (NCCL). For this very reason alone we _should_ actually suggest the affiliation with _ranks_. In addition this is what the user (GNN consumer) wants to call them.

File: cpp/include/cugraph/algorithms.hpp
Comment (aschaffer) [reply cpp/include/cugraph/algorithms.hpp:1542]: Fixed.

File: cpp/include/cugraph/algorithms.hpp
Comment (aschaffer) [reply cpp/include/cugraph/algorithms.hpp:1549]: Fixed and fixed.

File: cpp/src/sampling/nbr_sampling.cuh
Comment (aschaffer) [reply cpp/src/sampling/nbr_sampling.cuh:475]: So, what's the desired course of action here: _no_ documentation? some documentation...? I'll remove it for now.

File: cpp/src/sampling/nbr_sampling.cu
Comment (aschaffer) [reply cpp/src/sampling/nbr_sampling.cu:1]: Will do.

**Comment by aschaffer** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (ChuckHastings) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: OK.  I notice in the call to this function that the parameter passed as minor_indices is never used again.  Makes me wonder why it needs to be modified in place and maintained.

File: cpp/src/sampling/nbr_sampling.cuh
Comment (ChuckHastings) [reply cpp/src/sampling/nbr_sampling.cuh]: Generally I think the goal is to document functions in the place where they are defined.  If the function is implemented somewhere else then we don't want to replicate the documentation of the definition, since that leads to more potential for the documentation becoming inconsistent over time.  With our templating, we sometimes have a function defined in one place and implemented in multiple other places.  It just becomes a maintenance challenge if the same documentation is in multiple places.

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: It's part of the output that GNN wants.

File: cpp/src/sampling/nbr_sampling.cuh
Comment (aschaffer) [reply cpp/src/sampling/nbr_sampling.cuh]: Done.

**Review by seunghwak — APPROVED** [review]

**Review by kaatish — APPROVED** [review]

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (kaatish) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: @ChuckHastings gather_local_edges does not return minor_indices again.
This is what it returns :
`  return std::make_tuple(std::move(majors), std::move(minors), std::move(minor_gpu_ids));`
minor_indices is passed as a reference because it is in/out param.

File: cpp/tests/sampling/detail/mg_gather_utils.cu
Comment (kaatish) [inline cpp/tests/sampling/detail/mg_gather_utils.cu:244]: This needs to be reverted.

**Review by kaatish — CHANGES_REQUESTED** [review]

File: cpp/tests/sampling/detail/mg_gather_utils.cu
Comment (aschaffer) [reply cpp/tests/sampling/detail/mg_gather_utils.cu]: Done.

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (ChuckHastings) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: Yes.  So can we clarify what is the difference between the `minor_indices` that is the input/output parameter and the `minors` that is returned as part of the tuple.

Comment (seunghwak) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: Just a suggestion, mixing in/out parameters & return values might be a bit confusing, and can't we take `minor_indices` as an R-value return minor_indices as well? (instead of updating in-place).

Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: > can't we take minor_indices as an R-value return minor_indices as well

That would require an avoidable copy / move. The indices already exist, they just need to be filtered. You can pass them as `const&` but then in order to filter them you need a copy; i.e., extra storage + a move (at least; via RVO). We can avoid that, especially because this can be a large vector.

**Comment by aschaffer** [comment]
Note: Requires https://github.com/rapidsai/cugraph/pull/2110 merged to resolve current CI failures.

File: cpp/include/cugraph/detail/graph_functions.cuh
Comment (seunghwak) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: I am not sure I am getting what you're saying.

If we're taking `minor_indices` as an **R-value**, we can modify the variable and return the variable.
```
...
rmm::device_uvector<typename GraphViewType::edge_type>&& minor_indices,
...
{
  filter(minor_indices);
  ...
  return std::make_tuple(..., std::move(minor_indices), ...);
}
```

Am I missing something?

Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: @seunghwak , unless you meant passing `minor_indices` as rvalue and then returning it as part of the tuple. IOW, did you mean:
```
vector foo(vector&& minor_indices){
  // blah... modify minor_indices;
  return minor_indices;
}
```
That only involves moves. That's what you meant?

Comment (seunghwak) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: Yes!!! This wasn't possible when we were passing a pointer but now we're passing a device_uvector, and this is now possible.

Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh:185]: Okay. Let me do that.

Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh]: Done.

Comment (aschaffer) [reply cpp/include/cugraph/detail/graph_functions.cuh]: Also addressed the naming issue and corresponding signature and dox.

**Review by ChuckHastings — APPROVED** [review]

**Review by kaatish — APPROVED** [review]

File: cpp/include/cugraph/algorithms.hpp
Comment (rlratzel) [reply cpp/include/cugraph/algorithms.hpp:1571]: FWIW, I also found the word `rank` confusing in this context at first since it wasn't clear if it was intended to be a GPU rank for use by the sampling algo itself, or a rank used by the caller (ie. a client may have their own notion of rank independent of the MG rank used by this algo). I think a more general-purpose term like `label` or `id` not only works in this case, but can also be used as a more generic mechanism for associating results to a caller's seeds. Lastly, the GNN client will likely not be directly calling this API anyway (they'll be using python), so we can make something more application-specific for them in python if necessary.

**Review by rlratzel — APPROVED** [review]
My only feedback was wrt the `rank` naming discussion, but approving so it gets merged.

**Comment by BradReesWork** [comment]
@gpucibot merge

**Comment by Ethyling** [comment]
Removing ops-codeowners from the required reviews since it doesn't seem there are any file changes that we're responsible for.

**Comment by Ethyling** [comment]
@gpucibot merge

**Pull request merged**
